golang/session/util_buffer.go

```diff
@@ -0,0 +1,93 @@
+package session
+
+import (
+	"io"
+	"sync"
+)
+
+// buffer provides a linked list buffer for data exchange
+// between producer and consumer. Theoretically the buffer is
+// of unlimited capacity as it does no allocation of its own.
+type buffer struct {
```

Comment:
Hm, what's the upside of the linked list approach vs building on top of https://golang.org/pkg/sync/#Pool?

Reply:
I was also considering that this could be satisfied by a `sync.Map` but since there are some cases where they are slower and this was originally in the /x/crypto/ssh implementation (where most of this code came from), I figured I should leave it. I didn't think about a Pool but that's even harder to see how it would replace. Thoughts?

Reply:
I think what threw me off is I assumed this exists to reduce allocations, which is also the point of sync.Pool, but now that I read closer it's more about providing concurrent read/write access to chunks in a sequential way. Does that sound right?

Reply:
Yes I believe so. Instead of locking for one big buffer, a reader can read chunks off while there are no writes/eofs, and writes/eofs can only be done sequentially.

Reply:
Oh, sorry ... I was thinking chanList could be replaced by `sync.Map`, I was confused when I tried to make sense of that.